File: sensor/src/extraction/spawn-invokes.ts
/**
 * JS/TS `child_process` process-spawn detection and target folding (ADR 0048,
 * Lattice sensor correctness fix c/1 — first "call graph非可視の結合" index).
 *
 * `spawnSync(process.execPath, [BIN, ...args])` reaches `BIN` by starting a
 * brand-new OS process, not by calling a JS function — no `calls` edge can
 * ever represent it, so a spawn-driven test harness (the oracle shape:
 * `tests/orchestrate/helpers.mjs` spawning `bin/orchestrate-run.mjs`) showed
 * zero `affected` dependents even though changing the spawned file plainly
 * breaks the test. This module detects the five `child_process` launch
 * functions (`spawn`, `spawnSync`, `execFile`, `execFileSync`, `fork`) —
 * `exec`/`execSync` (shell-string parsing) are OUT OF SCOPE, see
 * `resolveChildProcessCallee`'s doc — and folds their target argument(s) to a
 * project-relative path using the same constant-folding engine as dynamic
 * `import()`/`require()` (dynamic-import.ts's `foldConstantExpr`).
 *
 * Detection REQUIRES the callee to be verified bound to `child_process` (a
 * static `import`/aliased/namespace member, or a simple
 * `const { spawnSync } = require('node:child_process')`) — a bare name match
 * on a user-defined `spawn()` must NOT produce an edge (mirrors fix (a)'s
 * no-name-match-fallback principle).
 */

import type { Node as SyntaxNode } from 'web-tree-sitter';
import * as posixPath from 'path';
import { getChildByField, getNodeText } from './tree-sitter-helpers';
import { foldConstantExpr } from './dynamic-import';

/**
 * Sentinel `referenceName` for a `fork(...)` call whose module argument could
 * NOT be statically folded. Mirrors `DYNAMIC_IMPORT_UNRESOLVED_MARKER`
 * (dynamic-import.ts): illegal as both a JS identifier and a file path
 * (angle brackets, colon), so it can never coincidentally name-match a real
 * file, and is pushed as a normal unresolved reference so it surfaces as
 * `status='failed'` rather than being silently dropped. `fork`'s target is
 * ALWAYS a JS module (unlike `spawn`/`exec*`, whose target is usually an
 * external binary) — so an unfoldable `fork` argument stays visible, while
 * an unfoldable `spawn`/`execFile*` argument is a silent skip (see
 * `resolveInvokesTargets`'s case (iv) — spawn's target domain is mostly
 * external commands, so "couldn't fold" carries no useful signal there).
 */
export const SPAWN_INVOKES_UNRESOLVED_MARKER = '<spawn:unresolved>';

const CHILD_PROCESS_SPAWN_FNS = new Set(['spawn', 'spawnSync', 'execFile', 'execFileSync', 'fork']);

function isChildProcessSpecifier(text: string): boolean {
  return text === 'child_process' || text === 'node:child_process';
}

/** Per-file table of local names verified bound to `child_process`'s launch functions. */
export interface ChildProcessBindings {
  /** local identifier -> canonical child_process function name (`spawn`, `fork`, ...) */
  readonly localFns: ReadonlyMap<string, string>;
  /** local identifiers bound as a `child_process` namespace (`import * as cp from 'child_process'`) */
  readonly namespaces: ReadonlySet<string>;
}

const EMPTY_BINDINGS: ChildProcessBindings = { localFns: new Map(), namespaces: new Set() };

/**
 * Scan the file's top-level (Program direct-child) statements for bindings of
 * `child_process`'s spawn family. Two binding shapes, per spec:
 *   - static `import` — named (`import { spawnSync } from 'child_process'`,
 *     alias-aware), or namespace (`import * as cp from 'node:child_process'`,
 *     later matched via `cp.spawnSync(...)`).
 *   - simple CommonJS destructure — `const { spawnSync } = require('node:child_process')`
 *     (module-level only; `let`/`var` excluded like dynamic-import.ts's
 *     module-level-const fold, since a reassignable binding's declaration-site
 *     value isn't necessarily the call-site value).
 * A default import isn't a meaningful binding shape for this module (no
 * useful default export) and is intentionally not recognized.
 */
export function collectChildProcessBindings(anyNodeInTree: SyntaxNode, source: string): ChildProcessBindings {
  const root = anyNodeInTree.tree.rootNode;
  const localFns = new Map<string, string>();
  const namespaces = new Set<string>();

  for (let i = 0; i < root.namedChildCount; i++) {
    let stmt = root.namedChild(i);
    if (!stmt) continue;

    if (stmt.type === 'import_statement') {
      const sourceField = getChildByField(stmt, 'source');
      const specifierText = sourceField ? getNodeText(sourceField, source).replace(/['"]/g, '') : '';
      if (!isChildProcessSpecifier(specifierText)) continue;
      const clause = stmt.namedChildren.find((c) => c?.type === 'import_clause');
      if (!clause) continue;
      for (const child of clause.namedChildren) {
        if (!child) continue;
        if (child.type === 'named_imports') {
          for (const spec of child.namedChildren) {
            if (!spec || spec.type !== 'import_specifier') continue;
            const nameNode = getChildByField(spec, 'name');
            const aliasNode = getChildByField(spec, 'alias');
            const imported = nameNode ? getNodeText(nameNode, source) : '';
            if (!CHILD_PROCESS_SPAWN_FNS.has(imported)) continue;
            const local = aliasNode ? getNodeText(aliasNode, source) : imported;
            if (local) localFns.set(local, imported);
          }
        } else if (child.type === 'namespace_import') {
          const idNode = child.namedChildren.find((c) => c?.type === 'identifier') ?? child.namedChild(0);
          const local = idNode ? getNodeText(idNode, source) : '';
          if (local) namespaces.add(local);
        }
      }
      continue;
    }

    if (stmt.type === 'lexical_declaration') {
      const kindNode = getChildByField(stmt, 'kind');
      if (kindNode?.type !== 'const') continue;
      for (let j = 0; j < stmt.namedChildCount; j++) {
        const declarator = stmt.namedChild(j);
        if (declarator?.type !== 'variable_declarator') continue;
        const nameNode = getChildByField(declarator, 'name');
        const valueNode = getChildByField(declarator, 'value');
        if (!nameNode || nameNode.type !== 'object_pattern' || !valueNode) continue;
        if (valueNode.type !== 'call_expression') continue;
        const fn = getChildByField(valueNode, 'function');
        if (!fn || fn.type !== 'identifier' || getNodeText(fn, source) !== 'require') continue;
        const argsNode = getChildByField(valueNode, 'arguments');
        const args = argsNode ? argsNode.namedChildren.filter((n): n is SyntaxNode => !!n) : [];
        if (args.length !== 1 || args[0]!.type !== 'string') continue;
        const fragment = args[0]!.namedChildren.find((c) => c?.type === 'string_fragment');
        const specifierText = fragment ? getNodeText(fragment, source) : '';
        if (!isChildProcessSpecifier(specifierText)) continue;

        for (let k = 0; k < nameNode.namedChildCount; k++) {
          const prop = nameNode.namedChild(k);
          if (!prop) continue;
          let imported = '';
          let local = '';
          if (prop.type === 'shorthand_property_identifier_pattern') {
            imported = getNodeText(prop, source);
            local = imported;
          } else if (prop.type === 'pair_pattern') {
            const keyNode = getChildByField(prop, 'key');
            const valNode = getChildByField(prop, 'value');
            imported = keyNode ? getNodeText(keyNode, source) : '';
            local = valNode ? getNodeText(valNode, source) : imported;
          } else {
            continue;
          }
          if (imported && local && CHILD_PROCESS_SPAWN_FNS.has(imported)) localFns.set(local, imported);
        }
      }
    }
  }

  return localFns.size === 0 && namespaces.size === 0 ? EMPTY_BINDINGS : { localFns, namespaces };
}

/**
 * True when `node` is `process.execPath` (a `member_expression`, not folded
 * by `foldConstantExpr` — its only member_expression shape is
 * `import.meta.dirname`). Distinguishing this from a general expression fold
 * is what lets `spawn(process.execPath, [...])` route into the array-element
 * folding branch (case (ii) below) instead of failing to fold argv[0].
 */
function isProcessExecPath(node: SyntaxNode, source: string): boolean {
  if (node.type !== 'member_expression') return false;
  const object = getChildByField(node, 'object');
  const property = getChildByField(node, 'property');
  return (
    !!object &&
    object.type === 'identifier' &&
    getNodeText(object, source) === 'process' &&
    !!property &&
    getNodeText(property, source) === 'execPath'
  );
}

/**
 * Given a `call_expression` node, return the canonical `child_process`
 * function name (`spawn`/`spawnSync`/`execFile`/`execFileSync`/`fork`) it
 * invokes, or null when the callee isn't verified bound to one. Two callee
 * shapes: a bare identifier bound via `localFns` (static named import or the
 * `require` destructure), or a `<namespace>.<fn>` member access where
 * `<namespace>` is a verified `import * as cp from 'child_process'` binding.
 *
 * `exec`/`execSync` are deliberately excluded — their single argument is a
 * whole shell command line (`"git status && ls"`), which requires shell
 * grammar parsing (quoting, pipes, `&&`, env vars) to extract a target file,
 * not the argv-array shape this module folds. Out of scope for this fix.
 */
export function resolveChildProcessCallee(
  node: SyntaxNode,
  source: string,
  bindings: ChildProcessBindings
): string | null {
  if (node.type !== 'call_expression') return null;
  const fn = getChildByField(node, 'function');
  if (!fn) return null;

  if (fn.type === 'identifier') {
    return bindings.localFns.get(getNodeText(fn, source)) ?? null;
  }

  if (fn.type === 'member_expression') {
    const object = getChildByField(fn, 'object');
    const property = getChildByField(fn, 'property');
    if (!object || object.type !== 'identifier' || !property) return null;
    if (!bindings.namespaces.has(getNodeText(object, source))) return null;
    const propName = getNodeText(property, source);
    return CHILD_PROCESS_SPAWN_FNS.has(propName) ? propName : null;
  }

  return null;
}

/**
 * Fold `node` to a constant string (`foldConstantExpr`), then — ONLY when the
 * result is a `./`/`../`-relative specifier — rewrite it into the same
 * project-relative, extension-included path space the resolver's exact-match
 * expects (import-resolver.ts's `invokes` branch does plain path equality,
 * no module resolution). A `join(ROOT, ...)`/`resolve(...)`-folded result is
 * already project-relative (dynamic-import.ts's `foldJoinResolve` produces
 * that form) and passes through untouched; a bare command name (`'git'`) has
 * no `./`/`../` prefix and also passes through untouched — the resolver's
 * exact match on a real file node is what turns "didn't fold to a real
 * file" into "no edge", not this function. A relative specifier that
 * resolves outside the project root is a fold failure (can't represent
 * "outside the project" as a project-relative path) — mirrors
 * `foldJoinResolve`'s same rule for `join`/`resolve` arithmetic.
 */
function foldToProjectPath(node: SyntaxNode, source: string, filePath: string): string | null {
  const folded = foldConstantExpr(node, source, filePath);
  if (folded === null) return null;
  if (!folded.startsWith('./') && !folded.startsWith('../')) return folded;

  const dir = posixPath.posix.dirname(filePath.replace(/\\/g, '/'));
  const base = dir === '.' ? '' : dir;
  const joined = base ? `${base}/${folded}` : folded;
  const normalized = posixPath.posix.normalize(joined);
  if (normalized === '..' || normalized.startsWith('../')) return null;
  return normalized;
}

/** One candidate `invokes` target: a folded path (real or not — the resolver decides) or the fork sentinel. */
export interface InvokesTarget {
  readonly referenceName: string;
}

/**
 * Fold `callNode`'s target argument(s) into zero or more `invokes` reference
 * names, per `canonicalFn`:
 *
 *  - `fork(modulePath, args?, options?)`: fold `modulePath` (argv[0] only —
 *    `fork`'s target is always a JS module, so it is ALWAYS pushed, using the
 *    unresolved sentinel on a folding failure rather than staying silent).
 *  - `spawn`/`spawnSync`/`execFile`/`execFileSync`(cmd, args?, options?):
 *      (i)   `cmd` folds to a path — push it. The resolver's exact-file-path
 *            match is the real filter: a project file resolves, an external
 *            command (`git`) or arbitrary string simply never matches a file
 *            node and silently stays unresolved (case iii from spec, "正しい
 *            挙動" — no special-casing needed here).
 *      (ii)  `cmd` is `process.execPath` (AST shape) or folds to the literal
 *            string `'node'` — argv[0] is the current runtime, so the REAL
 *            target is argv[1]'s array-literal elements. Each element is
 *            folded independently; a `...spread` or an unfoldable element is
 *            skipped WITHOUT abandoning the rest of the array.
 *      (iv)  `cmd` doesn't fold at all — skip entirely (no ref pushed).
 *            Unlike `fork`, spawn's target domain is mostly external
 *            commands, so an unfoldable argv[0] carries no signal that
 *            visibility would improve on.
 */
export function resolveInvokesTargets(
  canonicalFn: string,
  callNode: SyntaxNode,
  source: string,
  filePath: string
): InvokesTarget[] {
  const argsNode = getChildByField(callNode, 'arguments');
  if (!argsNode) return [];
  const args = argsNode.namedChildren.filter((n): n is SyntaxNode => !!n);
  if (args.length < 1) return [];

  if (canonicalFn === 'fork') {
    const folded = foldToProjectPath(args[0]!, source, filePath);
    return [{ referenceName: folded ?? SPAWN_INVOKES_UNRESOLVED_MARKER }];
  }

  const first = args[0]!;
  const foldedFirst = isProcessExecPath(first, source) ? null : foldToProjectPath(first, source, filePath);
  const isNodeSelfInvocation = isProcessExecPath(first, source) || foldedFirst === 'node';

  if (isNodeSelfInvocation) {
    const second = args[1];
    if (!second || second.type !== 'array') return [];
    const targets: InvokesTarget[] = [];
    for (const el of second.namedChildren) {
      if (!el || el.type === 'spread_element') continue; // 個別スキップ、配列全体は諦めない
      const foldedEl = foldToProjectPath(el, source, filePath);
      if (foldedEl !== null) targets.push({ referenceName: foldedEl });
    }
    return targets;
  }

  if (foldedFirst === null) return []; // case (iv): unfoldable argv[0] — skip, no ref
  return [{ referenceName: foldedFirst }]; // case (i)/(iii): let the resolver's exact-match decide
}

File: sensor/lattice-sensor-kernel/src/tsjs/spawn_invokes.rs
//! `child_process` spawn-family `invokes` edge indexing — a faithful Rust port
//! of src/extraction/spawn-invokes.ts (ADR 0048).
//!
//! `invokes` is an ADDITIONAL edge: the generic call-reference logic still
//! emits its normal `calls` ref to the local callee name.

use super::dynamic_import::fold_constant_expr;
use std::collections::{HashMap, HashSet};
use tree_sitter::Node;

/// Sentinel for a `fork(...)` whose module argument could not be folded.
/// Illegal as both a JS identifier and a file path, so it can never
/// coincidentally name-match a real file; pushed as a normal unresolved
/// reference so it surfaces as failed rather than being silently dropped.
pub const SPAWN_INVOKES_UNRESOLVED_MARKER: &str = "<spawn:unresolved>";

const SPAWN_FNS: [&str; 5] = ["spawn", "spawnSync", "execFile", "execFileSync", "fork"];

fn is_spawn_fn(name: &str) -> bool {
    SPAWN_FNS.contains(&name)
}

fn is_child_process_specifier(text: &str) -> bool {
    text == "child_process" || text == "node:child_process"
}

fn text<'a>(node: Node, src: &'a str) -> &'a str {
    &src[node.byte_range()]
}

/// Per-file table of local names verified bound to the spawn family.
#[derive(Default)]
pub struct ChildProcessBindings {
    /// local identifier -> canonical child_process function name
    local_fns: HashMap<String, String>,
    /// locals bound as a namespace (`import * as cp from 'child_process'`)
    namespaces: HashSet<String>,
}

impl ChildProcessBindings {
    pub fn is_empty(&self) -> bool {
        self.local_fns.is_empty() && self.namespaces.is_empty()
    }
}

/// Scan Program direct children for spawn-family bindings. Two shapes: static
/// `import` (named with alias, or namespace), and a module-level CommonJS
/// destructure (`const { spawnSync } = require('node:child_process')`).
/// `let`/`var` are excluded — a reassignable binding's declaration-site value
/// isn't necessarily the call-site value. A default import isn't a meaningful
/// binding shape for this module and is intentionally not recognized.
pub fn collect_child_process_bindings(any_node: Node, src: &str) -> ChildProcessBindings {
    let mut root = any_node;
    while let Some(parent) = root.parent() {
        root = parent;
    }
    let mut bindings = ChildProcessBindings::default();

    for i in 0..root.named_child_count() {
        let Some(stmt) = root.named_child(i) else { continue };

        if stmt.kind() == "import_statement" {
            let specifier = stmt
                .child_by_field_name("source")
                .map(|s| text(s, src).replace(['\'', '"'], ""))
                .unwrap_or_default();
            if !is_child_process_specifier(&specifier) {
                continue;
            }
            let Some(clause) = (0..stmt.named_child_count())
                .filter_map(|j| stmt.named_child(j))
                .find(|c| c.kind() == "import_clause")
            else {
                continue;
            };
            for j in 0..clause.named_child_count() {
                let Some(child) = clause.named_child(j) else { continue };
                match child.kind() {
                    "named_imports" => {
                        for k in 0..child.named_child_count() {
                            let Some(spec) = child.named_child(k) else { continue };
                            if spec.kind() != "import_specifier" {
                                continue;
                            }
                            let imported = spec
                                .child_by_field_name("name")
                                .map(|n| text(n, src).to_string())
                                .unwrap_or_default();
                            if !is_spawn_fn(&imported) {
                                continue;
                            }
                            let local = spec
                                .child_by_field_name("alias")
                                .map(|a| text(a, src).to_string())
                                .unwrap_or_else(|| imported.clone());
                            if !local.is_empty() {
                                bindings.local_fns.insert(local, imported);
                            }
                        }
                    }
                    "namespace_import" => {
                        let id = (0..child.named_child_count())
                            .filter_map(|k| child.named_child(k))
                            .find(|c| c.kind() == "identifier")
                            .or_else(|| child.named_child(0));
                        if let Some(id) = id {
                            let local = text(id, src).to_string();
                            if !local.is_empty() {
                                bindings.namespaces.insert(local);
                            }
                        }
                    }
                    _ => {}
                }
            }
            continue;
        }

        if stmt.kind() == "lexical_declaration" {
            if stmt.child_by_field_name("kind").map(|k| k.kind()) != Some("const") {
                continue;
            }
            for j in 0..stmt.named_child_count() {
                let Some(declarator) = stmt.named_child(j) else { continue };
                if declarator.kind() != "variable_declarator" {
                    continue;
                }
                let Some(name_node) = declarator.child_by_field_name("name") else { continue };
                let Some(value) = declarator.child_by_field_name("value") else { continue };
                if name_node.kind() != "object_pattern" || value.kind() != "call_expression" {
                    continue;
                }
                let Some(f) = value.child_by_field_name("function") else { continue };
                if f.kind() != "identifier" || text(f, src) != "require" {
                    continue;
                }
                let Some(args_node) = value.child_by_field_name("arguments") else { continue };
                let args: Vec<Node> = (0..args_node.named_child_count())
                    .filter_map(|k| args_node.named_child(k))
                    .collect();
                if args.len() != 1 || args[0].kind() != "string" {
                    continue;
                }
                let specifier = (0..args[0].named_child_count())
                    .filter_map(|k| args[0].named_child(k))
                    .find(|c| c.kind() == "string_fragment")
                    .map(|fr| text(fr, src).to_string())
                    .unwrap_or_default();
                if !is_child_process_specifier(&specifier) {
                    continue;
                }

                for k in 0..name_node.named_child_count() {
                    let Some(prop) = name_node.named_child(k) else { continue };
                    let (imported, local) = match prop.kind() {
                        "shorthand_property_identifier_pattern" => {
                            let n = text(prop, src).to_string();
                            (n.clone(), n)
                        }
                        "pair_pattern" => {
                            let imported = prop
                                .child_by_field_name("key")
                                .map(|n| text(n, src).to_string())
                                .unwrap_or_default();
                            let local = prop
                                .child_by_field_name("value")
                                .map(|n| text(n, src).to_string())
                                .unwrap_or_else(|| imported.clone());
                            (imported, local)
                        }
                        _ => continue,
                    };
                    if !imported.is_empty() && !local.is_empty() && is_spawn_fn(&imported) {
                        bindings.local_fns.insert(local, imported);
                    }
                }
            }
        }
    }

    bindings
}

/// True when `node` is `process.execPath` — not something `fold_constant_expr`
/// handles (its only member_expression shape is `import.meta.dirname`), and
/// recognizing it is what routes `spawn(process.execPath, [...])` into the
/// array-element branch instead of failing to fold argv[0].
fn is_process_exec_path(node: Node, src: &str) -> bool {
    if node.kind() != "member_expression" {
        return false;
    }
    let (Some(object), Some(property)) =
        (node.child_by_field_name("object"), node.child_by_field_name("property"))
    else {
        return false;
    };
    object.kind() == "identifier" && text(object, src) == "process" && text(property, src) == "execPath"
}

/// The canonical spawn-family function a call invokes, or None when the callee
/// isn't verified bound to one. `exec`/`execSync` are deliberately excluded —
/// their argument is a whole shell command line, which needs shell grammar
/// parsing rather than the argv-array shape this module folds.
pub fn resolve_child_process_callee(
    node: Node,
    src: &str,
    bindings: &ChildProcessBindings,
) -> Option<String> {
    if node.kind() != "call_expression" {
        return None;
    }
    let f = node.child_by_field_name("function")?;

    if f.kind() == "identifier" {
        return bindings.local_fns.get(text(f, src)).cloned();
    }

    if f.kind() == "member_expression" {
        let object = f.child_by_field_name("object")?;
        let property = f.child_by_field_name("property")?;
        if object.kind() != "identifier" || !bindings.namespaces.contains(text(object, src)) {
            return None;
        }
        let prop = text(property, src);
        return is_spawn_fn(prop).then(|| prop.to_string());
    }

    None
}

/// Fold to a constant, then — ONLY for a `./`/`../` relative specifier —
/// rewrite into the project-relative path space the resolver's exact match
/// expects. A join/resolve-folded result is already in that form and passes
/// through; a bare command name (`git`) also passes through untouched, since
/// the resolver's exact match on a real file node is what turns "didn't fold
/// to a real file" into "no edge". Escaping the project root is a failure.
fn fold_to_project_path(node: Node, src: &str, file_path: &str) -> Option<String> {
    let folded = fold_constant_expr(node, src, file_path)?;
    if !folded.starts_with("./") && !folded.starts_with("../") {
        return Some(folded);
    }
    let normalized_path = file_path.replace('\\', "/");
    let base = match normalized_path.rfind('/') {
        Some(i) => &normalized_path[..i],
        None => "",
    };
    let joined = if base.is_empty() { folded } else { format!("{base}/{folded}") };
    let normalized = super::dynamic_import::posix_normalize_for_spawn(&joined);
    if normalized == ".." || normalized.starts_with("../") {
        return None;
    }
    Some(normalized)
}

/// Fold a call's target argument(s) into zero or more `invokes` reference
/// names.
///
/// `fork(modulePath, …)` folds argv[0] and is ALWAYS pushed — its target is
/// always a JS module, so an unfoldable argument stays visible via the
/// sentinel. For `spawn`/`execFile*`: argv[0] that folds is pushed; argv[0]
/// that is `process.execPath` or the literal `"node"` means the real targets
/// are argv[1]'s array elements, each folded independently (a spread or an
/// unfoldable element is skipped without abandoning the rest); an argv[0] that
/// doesn't fold at all is skipped entirely, since spawn's target domain is
/// mostly external commands and "couldn't fold" carries no signal there.
pub fn resolve_invokes_targets(
    canonical_fn: &str,
    call: Node,
    src: &str,
    file_path: &str,
) -> Vec<String> {
    let Some(args_node) = call.child_by_field_name("arguments") else { return Vec::new() };
    let args: Vec<Node> = (0..args_node.named_child_count())
        .filter_map(|i| args_node.named_child(i))
        .collect();
    if args.is_empty() {
        return Vec::new();
    }

    if canonical_fn == "fork" {
        let folded = fold_to_project_path(args[0], src, file_path)
            .unwrap_or_else(|| SPAWN_INVOKES_UNRESOLVED_MARKER.to_string());
        return vec![folded];
    }

    let first = args[0];
    let exec_path = is_process_exec_path(first, src);
    let folded_first =
        if exec_path { None } else { fold_to_project_path(first, src, file_path) };
    let node_self_invocation = exec_path || folded_first.as_deref() == Some("node");

    if node_self_invocation {
        let Some(second) = args.get(1) else { return Vec::new() };
        if second.kind() != "array" {
            return Vec::new();
        }
        let mut targets = Vec::new();
        for i in 0..second.named_child_count() {
            let Some(el) = second.named_child(i) else { continue };
            if el.kind() == "spread_element" {
                continue; // skip the element, not the whole array
            }
            if let Some(folded) = fold_to_project_path(el, src, file_path) {
                targets.push(folded);
            }
        }
        return targets;
    }

    match folded_first {
        // Let the resolver's exact match decide whether this names a real file.
        Some(folded) => vec![folded],
        // Unfoldable argv[0] — skip, no ref.
        None => Vec::new(),
    }
}
